include/pybind11/detail/type_caster_base.h

```diff
@@ -939,6 +940,7 @@ template <typename type> class type_caster_base : public type_caster_generic {
     template <typename T, typename = enable_if_t<std::is_move_constructible<T>::value>>
     static auto make_move_constructor(const T *x) -> decltype(new T(std::move(*const_cast<T *>(x))), Constructor{}) {
```

Comment:
"INCORRECT"

I don't see any usage of x, actually - GCC is correct, this is ugly and wrong. It is used in the type expression, but it doesn't need to be - you could just as well use `std::declval` instead.

```suggestion
    static auto make_move_constructor(const T *x) -> decltype(&std::declval<T>())), Constructor{}) {
```

Reply:
(And, obviously, remove the "x"  and leave the parameter unnamed and remove the lines hiding warnings)

Reply:
It builds (after removing extra parenthesis) but fails to run. See below.

I only tried both suggestions together. I'd highly appreciate if we could reserve these side battles for separate PRs. It's fine with me if someone wants to spend the time going above and beyond, but in my position it's difficult to justify spending more time on a very localized, very small enhancement of existing code, only tangentially related to the bigger problem I need to solve.

```
Running tests in directory "/usr/local/google/home/rwgk/forked/pybind11/tests":
ImportError while loading conftest '/usr/local/google/home/rwgk/forked/pybind11/tests/conftest.py'.
conftest.py:19: in <module>
    import pybind11_tests  # noqa: F401
E   ImportError: return_value_policy = copy, but type MyEnum is non-copyable!
```

Here is the diff that led to the error above, in case I made a small mistake and someone wants to try it more:
```
diff --git a/include/pybind11/detail/type_caster_base.h b/include/pybind11/detail/type_caster_base.h
index a3ce4204..e84e8d99 100644
--- a/include/pybind11/detail/type_caster_base.h
+++ b/include/pybind11/detail/type_caster_base.h
@@ -929,18 +929,14 @@ protected:
     /* Only enabled when the types are {copy,move}-constructible *and* when the type
        does not have a private operator new implementation. */
     template <typename T, typename = enable_if_t<is_copy_constructible<T>::value>>
-    static auto make_copy_constructor(const T *x) -> decltype(new T(*x), Constructor{}) {
-        PYBIND11_WORKAROUND_INCORRECT_MSVC_C4100(x);
-        PYBIND11_WORKAROUND_INCORRECT_OLD_GCC_UNUSED_BUT_SET_PARAMETER(x);
+    static auto make_copy_constructor(const T *) -> decltype(&std::declval<T>(), Constructor{}) {
         return [](const void *arg) -> void * {
             return new T(*reinterpret_cast<const T *>(arg));
         };
     }
 
     template <typename T, typename = enable_if_t<std::is_move_constructible<T>::value>>
-    static auto make_move_constructor(const T *x) -> decltype(new T(std::move(*const_cast<T *>(x))), Constructor{}) {
-        PYBIND11_WORKAROUND_INCORRECT_MSVC_C4100(x);
-        PYBIND11_WORKAROUND_INCORRECT_OLD_GCC_UNUSED_BUT_SET_PARAMETER(x);
+    static auto make_move_constructor(const T *) -> decltype(&std::declval<T>(), Constructor{}) {
         return [](const void *arg) -> void * {
             return new T(std::move(*const_cast<T *>(reinterpret_cast<const T *>(arg))));
         };
```

Reply:
The thing you are replacing is different in the two cases, why would you expect it to become the same expression? `new T(*x)` is not the same as `new T(std::move(*const_cast<T *>(x))`. If it makes you feel better, I did exactly the same mistake about 6 times in a row trying to understand why this didn't work. :) 

The good news is that by doing it so many times, I have a better idea of what's going on here and was able to simplify the expression a bit - it's easier and more natural when you can make any type you want via declval, rather than having to cast around a preset value.

Reply:
Ahh, `gh pr checkout 3164` failed, so I modified and pushed to master without noticing it. Reverted, will reset locally and apply to this branch.

Reply:
I wondered why git apply didn't work...